### ISSUE TITLE
Possibly expensive SQL Query on determining files to process

### ISSUE THREAD

**Issue by kriegerse** [GDATASoftwareAG/nextcloud-gdata-antivirus#144]
# Description

Hey it's me again :-D

I've recognized a high load on my Maria DB Instance. It took me a while to identify what cause the load but the MariaDB slow log is very clear about the SQL Statement and I suspect to trace it down to the GData VaaS App.

It looks like the GData VaaS makes an uncomfortable SQL Query to the Database not performing well.

# Reproduce

- Install Nextcloud 
- Install gdatavaas App
- Possibly have a lot of files + folder in your nextcloud and/or oc_filecache table - I count:
  -  `~293491` in my nc/oc data dir (overall)
     - `~27351`  in my (sinlge) userdir  
     - `~265999` in my appdata folder (including previews)
  - `~ 349134`  in oc_filecache table

- Hardware simple NAS with:
  - Intel(R) Celeron(R) J4105 CPU @ 1.50GHz
  - 32GB RAM

For sure no enterprise grade hardware but also no numbers of files like enterprises have.

# Log(s) & Analysis

* MariaDB Slow Log - I see a couple of them in my slow.log but average run of the query is always around ~800 seconds (~14 Minutes)
```
# Time: 241002 18:45:26
# User@Host: oc_XXXXXX[oc_XXXXXX] @  [XXX.XXX.XXX.XXX]
# Thread_id: 32250  Schema: XXX-owncloud  QC_hit: No
# Query_time: 908.723772  Lock_time: 0.000139  Rows_sent: 10000  Rows_examined: 505241
# Rows_affected: 0  Bytes_sent: 120089
SET timestamp=1727894726;
SELECT `f`.`fileid` FROM `oc_filecache` `f` LEFT JOIN `oc_systemtag_object_mapping` `o` ON `f`.`fileid` = CAST(o.objectid AS UNSIGNED) LEFT JOIN `oc_mimetypes` `m` ON `f`.`mimetype` = `m`.`id` WHERE ((`o`.`systemtagid` NOT IN (248, 249, 251, 250, 252)) OR (`o`.`systemtagid` IS NULL)) AND (`m`.`mimetype` NOT LIKE '%unix-directory%') AND ((`f`.`path` LIKE 'files/%') OR (`f`.`path` LIKE '__groupfolders/%')) ORDER BY `f`.`fileid` DESC LIMIT 10000;
```

* MariaDB - Explain: about the statement
```
MariaDB [XXX-owncloud]> EXPLAIN SELECT `f`.`fileid` FROM `oc_filecache` `f` LEFT JOIN `oc_systemtag_object_mapping` `o` ON `f`.`fileid` = CAST(o.objectid AS UNSIGNED) LEFT JOIN `oc_mimetypes` `m` ON `f`.`mimetype` = `m`.`id` WHERE ((`o`.`systemtagid` NOT IN (248, 249, 251, 250, 252)) OR (`o`.`systemtagid` IS NULL)) AND (`m`.`mimetype` NOT LIKE '%unix-directory%') AND ((`f`.`path` LIKE 'files/%') OR (`f`.`path` LIKE '__groupfolders/%')) ORDER BY `f`.`fileid` DESC LIMIT 10000;
+------+-------------+-------+--------+---------------+---------+---------+----------------------------+--------+--------------------------------------------------------------+
| id   | select_type | table | type   | possible_keys | key     | key_len | ref                        | rows   | Extra                                                        |
+------+-------------+-------+--------+---------------+---------+---------+----------------------------+--------+--------------------------------------------------------------+
|    1 | SIMPLE      | f     | ALL    | NULL          | NULL    | NULL    | NULL                       | 118467 | Using where; Using temporary; Using filesort                 |
|    1 | SIMPLE      | m     | eq_ref | PRIMARY       | PRIMARY | 8       | XXX-owncloud.f.mimetype    | 1      | Using where                                                  |
|    1 | SIMPLE      | o     | index  | NULL          | PRIMARY | 524     | NULL                       | 7758   | Using where; Using index; Using join buffer (flat, BNL join) |
+------+-------------+-------+--------+---------------+---------+---------+----------------------------+--------+--------------------------------------------------------------+
3 rows in set (0.001 sec)
```
Not an MariaDB/SQL Expert but it looks like for the first part of the query with `oc_filecache` table no index can be utilized and endup in a expensive full table scan

* Simple Select of first `oc_filecache` is able to utilize index
```
MariaDB [XXX-owncloud]> EXPLAIN SELECT `f`.`fileid` FROM `oc_filecache` `f` ;
+------+-------------+-------+-------+---------------+----------+---------+------+--------+-------------+
| id   | select_type | table | type  | possible_keys | key      | key_len | ref  | rows   | Extra       |
+------+-------------+-------+-------+---------------+----------+---------+------+--------+-------------+
|    1 | SIMPLE      | f     | index | NULL          | fs_mtime | 8       | NULL | 118467 | Using index |
+------+-------------+-------+-------+---------------+----------+---------+------+--------+-------------+
1 row in set (0.000 sec)
```

* The tag IDs (248, 249, 251, 250, 252) seems to belong to GData VaaS to manage states
```
MariaDB [XXX-owncloud]> SELECT * FROM `oc_systemtag` WHERE `id` IN (248, 249, 251, 250, 252) ORDER BY id;
+-----+------------+------------+----------+
| id  | name       | visibility | editable |
+-----+------------+------------+----------+
| 248 | Unscanned  |          1 |        0 |
| 249 | Malicious  |          1 |        0 |
| 250 | Pup        |          1 |        0 |
| 251 | Clean      |          1 |        0 |
| 252 | Won't scan |          1 |        0 |
+-----+------------+------------+----------+
5 rows in set (0.001 sec)
``` 

## Database Stats

Some statistic data might help for the issue 

* Count Rows on `oc_filecache` (f)
```
MariaDB [XXX-owncloud]> SELECT COUNT(`f`.`fileid`) FROM `oc_filecache` `f` ;
+---------------------+
| COUNT(`f`.`fileid`) |
+---------------------+
|              349134 |
+---------------------+
1 row in set (0.121 sec)
```

* Count Rows on `oc_systemtag_object_mapping` (o)
```
MariaDB [XXX-owncloud]> SELECT COUNT(*) FROM `oc_systemtag_object_mapping`  ;
+----------+
| COUNT(*) |
+----------+
|   119045 |
+----------+
```

* Count Rows on `oc_mimetypes` (m)
```
MariaDB [XXX-owncloud]> SELECT COUNT(*) FROM `oc_mimetypes`  ;
+----------+
| COUNT(*) |
+----------+
|      184 |
+----------+
1 row in set (0.000 sec)
```

## Available Indices

* `oc_filecache`
```
MariaDB [XXX-owncloud]> show index from oc_filecache;
+--------------+------------+--------------------------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+
| Table        | Non_unique | Key_name                 | Seq_in_index | Column_name | Collation | Cardinality | Sub_part | Packed | Null | Index_type | Comment | Index_comment | Ignored |
+--------------+------------+--------------------------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+
| oc_filecache |          0 | PRIMARY                  |            1 | fileid      | A         |      118467 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          0 | fs_storage_path_hash     |            1 | storage     | A         |          12 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          0 | fs_storage_path_hash     |            2 | path_hash   | A         |      118467 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | fs_parent_name_hash      |            1 | parent      | A         |       31323 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | fs_parent_name_hash      |            2 | name        | A         |      118455 |     NULL | NULL   | YES  | BTREE      |         |               | NO      |
| oc_filecache |          1 | fs_storage_mimetype      |            1 | storage     | A         |          12 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | fs_storage_mimetype      |            2 | mimetype    | A         |         250 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | fs_storage_mimepart      |            1 | storage     | A         |          12 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | fs_storage_mimepart      |            2 | mimepart    | A         |          38 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | fs_storage_size          |            1 | storage     | A         |          12 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | fs_storage_size          |            2 | size        | A         |       70369 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | fs_storage_size          |            3 | fileid      | A         |      118467 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | fs_mtime                 |            1 | mtime       | A         |       42210 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | fs_size                  |            1 | size        | A         |       35679 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | fs_id_storage_size       |            1 | fileid      | A         |      118467 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | fs_id_storage_size       |            2 | storage     | A         |      118467 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | fs_id_storage_size       |            3 | size        | A         |      118467 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | fs_parent                |            1 | parent      | A         |      118467 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | memories_parent_mimetype |            1 | parent      | A         |      118467 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | memories_parent_mimetype |            2 | mimetype    | A         |      118467 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | memories_name_hash       |            1 | name        | A         |       29616 |     NULL | NULL   | YES  | BTREE      |         |               | NO      |
| oc_filecache |          1 | fs_storage_path_prefix   |            1 | storage     | A         |           9 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_filecache |          1 | fs_storage_path_prefix   |            2 | path        | A         |      118467 |       64 | NULL   | YES  | BTREE      |         |               | NO      |
+--------------+------------+--------------------------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+
23 rows in set (0.001 sec)
```

As other apps (memories) maintain there own indices, this might be also a solution for GData VaaS - if some is missing/can not utilized

* `oc_mimetypes`
```
MariaDB [XXX-owncloud]> show index from oc_mimetypes;
+--------------+------------+-------------------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+
| Table        | Non_unique | Key_name          | Seq_in_index | Column_name | Collation | Cardinality | Sub_part | Packed | Null | Index_type | Comment | Index_comment | Ignored |
+--------------+------------+-------------------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+
| oc_mimetypes |          0 | PRIMARY           |            1 | id          | A         |         171 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_mimetypes |          0 | mimetype_id_index |            1 | mimetype    | A         |         171 |     NULL | NULL   |      | BTREE      |         |               | NO      |
+--------------+------------+-------------------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+
2 rows in set (0.001 sec)
```

* `oc_systemtag_object_mapping`
```
MariaDB [XXX-owncloud]> show index from oc_systemtag_object_mapping;
+-----------------------------+------------+---------------------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+
| Table                       | Non_unique | Key_name            | Seq_in_index | Column_name | Collation | Cardinality | Sub_part | Packed | Null | Index_type | Comment | Index_comment | Ignored |
+-----------------------------+------------+---------------------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+
| oc_systemtag_object_mapping |          0 | PRIMARY             |            1 | objecttype  | A         |           1 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_systemtag_object_mapping |          0 | PRIMARY             |            2 | objectid    | A         |        7750 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_systemtag_object_mapping |          0 | PRIMARY             |            3 | systemtagid | A         |        7758 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_systemtag_object_mapping |          1 | systag_by_tagid     |            1 | systemtagid | A         |          19 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_systemtag_object_mapping |          1 | systag_by_tagid     |            2 | objecttype  | A         |          29 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_systemtag_object_mapping |          1 | memories_type_tagid |            1 | objecttype  | A         |           0 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_systemtag_object_mapping |          1 | memories_type_tagid |            2 | systemtagid | A         |          17 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| oc_systemtag_object_mapping |          1 | systag_by_objectid  |            1 | objectid    | A         |        2586 |     NULL | NULL   |      | BTREE      |         |               | NO      |
+-----------------------------+------------+---------------------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+
8 rows in set (0.000 sec)
```

# Expected behavior

* GData VaaS performs well with SQL queries - the SQL Statement mentioned shouldn't run minutes also with respect to enterprise grade instances

# Versions

- Nextcloud Hub 8 [29.0.7]
- Gdatavaas [29.4.3]
- MariaDB [10.11.9]

> Sorry for this possibly hard nut

**Comment by lennartdohmann** [comment]
Hi, welcome back :D Thanks for the detailed description and all the information. This will certainly help a lot. In any case, I agree that the app should be more performant than it seems to be now, even with so many files. Due to a public holiday tomorrow and the following bridge day (for most people), this will probably have to wait a few days. But I'll get straight to have a look on it on Monday.

**Comment by pstadermann** [comment]
It appears that we don't have to add new columns or indices. The current query is not filtering on oc_systemtag_object_mapping.objecttype. The composite index (objecttype, objectid, systemtagid) is not used.

If you find the time, it would be great to know, if the new query is faster than 1s on your system.

We will release the fix in the coming days.

## Test data

```
SELECT COUNT(*) FROM oc_filecache;
200470

SELECT COUNT(*) FROM oc_systemtag_object_mapping;
200044
```

## Current query

```sql
SELECT
    `f`.`fileid`
FROM
    `oc_filecache` `f`
    LEFT JOIN `oc_systemtag_object_mapping` `o` ON `f`.`fileid` = CAST(o.objectid AS UNSIGNED)
    LEFT JOIN `oc_mimetypes` `m` ON `f`.`mimetype` = `m`.`id`
WHERE
    (
        (
            `o`.`systemtagid` NOT IN (248, 249, 251, 250, 252)
        )
        OR (`o`.`systemtagid` IS NULL)
    )
    AND (`m`.`mimetype` NOT LIKE '%unix-directory%')
    AND (
        (`f`.`path` LIKE 'files/%')
        OR (`f`.`path` LIKE '__groupfolders/%')
    )
ORDER BY
    `f`.`fileid` DESC
LIMIT
    10000;
```

Runtime: Aborted after 18min

EXPLAIN:

| id | select\_type | table | type | possible\_keys | key | key\_len | ref | rows | filtered | Extra |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 1 | SIMPLE | f | ALL | null | null | null | null | 184619 | 100 | Using where; Using temporary; Using filesort |
| 1 | SIMPLE | m | eq\_ref | PRIMARY | PRIMARY | 8 | nextcloud.f.mimetype | 1 | 100 | Using where |
| 1 | SIMPLE | o | index | null | PRIMARY | 524 | null | 191839 | 100 | Using where; Using index; Using join buffer \(flat, BNL join\) |


## New query

```sql
SELECT
    `f`.`fileid`
FROM
    `oc_filecache` `f`
    LEFT JOIN `oc_systemtag_object_mapping` `o` ON `f`.`fileid` = CAST(o.objectid AS UNSIGNED)
    LEFT JOIN `oc_mimetypes` `m` ON `f`.`mimetype` = `m`.`id`
WHERE
    (
        (
            `o`.`systemtagid` NOT IN (248, 249, 251, 250, 252)
        )
        OR (`o`.`systemtagid` IS NULL)
    )
    AND (o.objecttype = 'files')
    AND (`m`.`mimetype` NOT LIKE '%unix-directory%')
    AND (
        (`f`.`path` LIKE 'files/%')
        OR (`f`.`path` LIKE '__groupfolders/%')
    )
ORDER BY
    `f`.`fileid` DESC
LIMIT
    10000;
```

Runtime: 800ms

EXPLAIN:

| id | select\_type | table | type | possible\_keys | key | key\_len | ref | rows | filtered | Extra |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 1 | SIMPLE | o | ref | PRIMARY,systag\_by\_tagid | PRIMARY | 258 | const | 95919 | 50 | Using where; Using index; Using temporary; Using filesort |
| 1 | SIMPLE | f | eq\_ref | PRIMARY,fs\_id\_storage\_size | PRIMARY | 8 | func | 1 | 100 | Using where |
| 1 | SIMPLE | m | eq\_ref | PRIMARY | PRIMARY | 8 | nextcloud.f.mimetype | 1 | 100 | Using where |

**Comment by pstadermann** [comment]
Still got some issues. The SQL query is not correct. Will update once I get it running.

**Comment by lennartdohmann** [comment]
Should be solved with e26678369d35d270079ffec53a8e3db459210e09 in the next release.

**Comment by kriegerse** [comment]
Can confirm the fix - didn't see any entries anymore in the slow log for the problematic query after latest release. 
again - thanks for fixing  :+1:
